src/threadpool.c

```diff
@@ -344,13 +375,71 @@ static void uv__queue_done(struct uv__work* w, int err) {
 }
 
 
+int uv__queue_no_after_work(uv_work_t* req, uv_work_cb work_cb) {
+  /* Don't use uv__req_init() since we're not increasing the active req count on
+   * any given uv_loop_t. */
+  UV_REQ_INIT(req, UV_WORK);
+  /* TODO(trevnorris): It's not necessary to set uv_loop_t when there's no
+   * after_work_cb. Should it still be assigned here? */
+  req->work_cb = work_cb;
+  /* Queue'ing without uv__queue_done will cause the req to be removed from the
+   * queue before calling the uv_work_cb. */
+  uv__work_submit(NULL,
+                  &req->work_req,
+                  UV__WORK_CPU,
+                  uv__queue_work,
+                  NULL);
+  return 0;
+}
+
+
+int uv__queue_no_work_cb(uv_loop_t* loop,
+                         uv_work_t* req,
+                         uv_after_work_cb after_work_cb) {
+  struct uv__work* w = &req->work_req;
+  int r;
+
+  uv__req_init(loop, req, UV_WORK);
+  req->loop = loop;
+  /* It's important this is set to NULL since it's used by uv__work_cancel()
+   * to identify that this uv_work_t has been queued without a uv_work_cb.
+   * Otherwise calling uv_cancel() will return UV_EBUSY. */
+  req->work_cb = NULL;
+  req->after_work_cb = after_work_cb;
+  /* Not calling into uv__work_submit since we don't need post() called. So
+   * set a few important fields here. */
+  w->loop = loop;
+  w->work = NULL;
+  w->done = uv__queue_done;
+
+  /* Insert this directly into the event loop's queue for execution and call
+   * to be executed. Since we're accessing the event loop off its thread we
+   * have to trust the user that the event loop is still valid. */
+  uv_mutex_lock(&loop->wq_mutex);
+  QUEUE_INSERT_TAIL(&loop->wq, &w->wq);
+  r = uv_async_send(&loop->wq_async);
+  uv_mutex_unlock(&loop->wq_mutex);
+
+  return r;
+}
+
+
 int uv_queue_work(uv_loop_t* loop,
                   uv_work_t* req,
                   uv_work_cb work_cb,
                   uv_after_work_cb after_work_cb) {
-  if (work_cb == NULL)
+  if (work_cb == NULL && after_work_cb == NULL)
     return UV_EINVAL;
 
+  if (after_work_cb == NULL)
+    return uv__queue_no_after_work(req, work_cb);
```

Comment:
I feel this makes for an odd API. It'd be more coherent to follow the loop-less uv_fs_blah() functions, where you can invoke them with `loop == NULL`.

Enforcing that `after_work_cb == NULL` makes sense, I think. There's no "home" thread to run the callback on.

Reply:
There are 3 ways this can be invoked.
```c
uv_queue_work(NULL, req, work_cb, NULL);
uv_queue_work(loop, req, NULL, after_work_cb);
uv_queue_work(loop, req, work_cb, after_work_cb);
```
If they do the following:
```c
uv_queue_work(loop, req, work_cb, NULL);
```
then `loop` will just be ignored like it would be for the `uv_fs_*()` APIs when you don't pass in a cb.

Are you saying we should do something like return `UV_EINVAL` if they call it without a `uv_after_work_cb` but do pass in `loop`?

Reply:
Added enforcement of `loop == NULL` if `after_work_cb == NULL`.